### PR TITLE
Add explicit justification of COMMIT and CONVERGE

### PR DESCRIPTION
**Pull request by ranchalp**
This PR Adds explicit justification, to be merged after #35 .

`quorumState` needs to store signatures (hard requirement). I also added a bitfield per tipsetID so that it easily identifies (and sorts by weight) the signers of a value (soft requirement). 

Additionally, `quorumState` needs to store messages that justify a value, for the case when a message triggers a value for the next step but the actual values that justify that message (and the message of the next step) have not been received outside the justification (hard requirement). An example is receiving a COMMIT with PREPARE in the COMMIT's evidence, but not having received the individual PREPAREs themselves. The participant may need then to use the same justification from the COMMIT directly in the CONVERGE.

TODO: - stop adding signatures once enough power has been found to reach the threshold.  
EDIT: the TODO is addressed in #50

### CODE REVIEW

File: f3/granite.go
Comment (anorth) [inline f3/granite.go:695]: It'll probably be easier to understand if this just a `map[ActorID]{}` here. Or `map[int]`, one of the problems is I'm not sure what the values represent. `Bitfield` just has that internally anyway. We can build a Bitfield for serialization when necessary.

File: f3/granite.go
Comment (anorth) [inline f3/granite.go:53]: I think either `Evidence` or `Justification` would be better names here. It's always aggregated. Calling it `justification` would align with all the uses of that word throughout the code.

**Review by anorth — COMMENTED** [review]
I've just skimmed this, and will come back to review properly once all the upstream PRs are landed. I understand the need abstractly to store the signatures of each value that must be subsequently used to create an evidence packet when the quorum is reached, and quorum state is probably a reasonable place.

I don't understand or am not convinced about the others, but they might make more sense as we iterate toward a clear expression here. I do get that if a node receives COMMIT with evidence, it will need to send its own COMMIT and copy the same evidence, but that doesn't create a need to store it. I don't think evidence ever propagates further than the immediate next phase.

File: f3/granite.go
Comment (ranchalp) [reply f3/granite.go]: Used `Justification` in the end

File: f3/granite.go
Comment (ranchalp) [reply f3/granite.go]: used `map[ActorID]struct{}`

**Comment by ranchalp** [comment]
I addressed almost all comments (simultaneously with rebasing). I realize maybe the first (new) commit of this PR might be a bit happy. I can split it into parts again.

I think you are probably right about not having to store every single value, but I will think of a smarter way of doing this probably tomorrow, since quorumState should not necessarily know which of the values will be used if there is more than one perhaps? perhaps not, I will go through the pseudocode tomorrow and update here. Will ping!

**Comment by ranchalp** [comment]
> I do get that if a node receives COMMIT with evidence, it will need to send its own COMMIT and copy the same evidence, but that doesn't create a need to store it. I don't think evidence ever propagates further than the immediate next phase.

Actually, Given the way in which we chose the proposal for the next round, we do not know which value will be chosen as the proposal for the next round. Any of the chose values might only be justified by the explicit justification with which it was sent, and not with locally received PREPAREs. Therefore, at least for the next round, we need to store one justification for each valid COMMIT with a non-zero value received. The map of `justifiedMessages` can be cleaned after sending the CONVERGE of the immediately following round yes (concretely right before returning at `beginConverge`), but until that is executed we need to store the justifiedMessages received (one per non-zero value as mentioned).

Not sure if it is an overkill to cleanup the map after sending the CONVERGE, would you like me to add that line? @anorth

**Comment by ranchalp** [comment]
@Kubuxu Same recommendation to just check files changed (and probably for all PRs of the chain)

File: f3/granite.go
Comment (Kubuxu) [inline f3/granite.go:106]: Keep this as Phase and use `.String()` internally.

File: f3/granite.go
Comment (Kubuxu) [inline f3/granite.go:90]: Keep this as Phase

File: f3/granite.go
Comment (Kubuxu) [reply f3/granite.go:106]: We will probably switch to encoding the Phase as int in the future (if I were to guess).

File: f3/granite.go
Comment (Kubuxu) [inline f3/granite.go:374]: My initial reaction was that this is hard to read. I will elaborate.

Comment (Kubuxu) [reply f3/granite.go:374]: My suggestion would be to use errors instead of logs (no error = justified).

File: f3/granite.go
Comment (Kubuxu) [inline f3/granite.go:394]: using `if` instead of the `else if` when the previous statement returns would also make it easier to read

File: f3/granite.go
Comment (Kubuxu) [inline f3/granite.go:85]: Justification, in general, has to be exactly the same as the signed message it is based on. This means that the GMessage struct is:
```go
type GMessage struct {
     Sender ActorID
     Current SignedMessage
     CurrentSig Signature
     Ticket Ticket
     Justification SignedMessage
     JustificationSignature Signature
}
```

where SignedMessage contains

File: f3/granite.go
Comment (Kubuxu) [inline f3/granite.go:531]: This should disappear if we adopt the GMessage changes I suggested.

File: f3/granite.go
Comment (ranchalp) [reply f3/granite.go:90]: Done.

File: f3/granite.go
Comment (ranchalp) [reply f3/granite.go:394]: Done.

File: f3/granite.go
Comment (ranchalp) [reply f3/granite.go:374]: Done, but I swapped your suggestion: because some conditions of semantic validation do not require (or have) any actual aggregate to verify, I instead have the `isJustified` perform semantic validation and call a new function `verifyJustification` at the end (if it has not returned with nil or error already before).

File: f3/granite.go
Comment (ranchalp) [reply f3/granite.go:531]: I am not quite sure how this would "disappear"? The justification still needs to be created and the signatures aggregated.

File: f3/granite.go
Comment (ranchalp) [reply f3/granite.go:85]: Makes sense. I instead made `Justification` have two fields `JustificationSignature JustificationSignature` and `Payload SignedMessage`, so that it is easy to carry both around (many times they go together).

**Comment by ranchalp** [comment]
Thanks @Kubuxu , just addressed all the comments.

File: f3/granite.go
Comment (Kubuxu) [inline f3/granite.go:391]: Please check the error on the outside and error out here. RLE+ has a bit of a sharp edge where it is easy to express "all bits between 0 and 2^64-1" which makes ForEach equivalent to an infinite loop. This is why many methods on it take `max` parameters (if the fact that bit-by-bit iteration happens is unclear). Another way to guard against it is to compare `rle.Count()` against some constant.
We might have to define a constant of `MaxQuorum = 10e6` or something like that (10e6 is arbitrarily large).

File: f3/granite.go
Comment (Kubuxu) [inline f3/granite.go:379]: Probably combine this into one iteration, as well as range checks.

File: f3/granite.go
Comment (Kubuxu) [reply f3/granite.go:533]: by that, I meant that the SignedMessage could have been stored in the prevRoundState but it is not that important

File: f3/granite.go
Comment (Kubuxu) [inline f3/granite.go:890]: Please don't introduce new panics, let's use error propagation

Comment (ranchalp) [reply f3/granite.go:890]: Separate PR it is then.

File: f3/granite.go
Comment (ranchalp) [reply f3/granite.go:375]: Good point, I changed the parameter to accept the justification only.

File: f3/granite.go
Comment (ranchalp) [reply f3/granite.go]: I changed all `ForEach` to check the bit against the powertable entries and return an error if too big (and propagated the error on the outside). Not sure if anything else is required for the `ForEach` given how we use it.

File: f3/granite.go
Comment (ranchalp) [reply f3/granite.go:382]: See above comment.

**Comment by ranchalp** [comment]
Thanks for the comments! addressed

**Review by Kubuxu — APPROVED** [review]
Let's get this merged to unblock (and so it doesn't get out of date).

**Pull request merged**
